Better operator string represenation (take 2)

johanandren opened the pull request:
Based on @jrudolph s great idea in #29874 (so supersedes that PR)

Instead of adding user provided lambda source location in the name attribute this PR introduces a new Attribute with source location. The default toString of `GraphStageLogic` has then been updated to include that information if available and relevant debug/error and stream snapshot labels for the operator have been updated to use the toString of `GraphStageLogic`. 

Example of preStart exception from one of the tests for example:
```
Error during preStart in [akka.stream.impl.SetupFlowStage$$anon$2-setup(akka.stream.scaladsl-FromMaterializationSpec.scala:141)]: boom
```

The format is `[logic instance class name]-[name attribute]([source location of lambda])`

johanandren committed:
New attribute with source location information introduced and added to stages that takes lambdas

johanandren committed:
Better default toString for GraphStageLogic including source location where possible

... and use that for debugging, errors and stream snapshots

johanandren committed:
Some more operators that take lambdas and tests fixed to pass

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala:
```diff
@@ -306,7 +306,7 @@ import akka.stream.stage._
         logic.preStart()
       } catch {
         case NonFatal(e) =>
-          log.error(e, "Error during preStart in [{}]: {}", logic.originalStage.getOrElse(logic), e.getMessage)
```
not sure why we ever thought that it was better to log the GraphStage rather than the actual logic here, feels like this is much better coming from the logic allowing a custom toString on it potentially provide logic state etc.

johanandren commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FromMaterializationSpec.scala:
```diff
@@ -46,7 +46,7 @@ class FromMaterializerSpec extends StreamSpec {
         .named("my-name")
 
-      source.runWith(Sink.head).futureValue shouldBe Some("my-name")
+      source.runWith(Sink.head).futureValue shouldBe Some("setup-my-name")
```
The name attribute is special and is concatenated, so since I added a default name "setup" it now becomes "setup-my-name" when adding "my-name"

johanandren commented on akka-stream-tests/src/test/scala/akka/stream/impl/GraphStageLogicSpec.scala:
```diff
@@ -282,7 +282,7 @@ class GraphStageLogicSpec extends StreamSpec with GraphInterpreterSpecKit with S
           .runWith(Sink.ignore)
       }
-      ex.getMessage should startWith("No handler defined in stage [stage-name] for in port [in")
+      ex.getMessage should startWith("No handler defined in stage [").and(include("] for in port [in"))
```
stage name is now to tricky to be worth checking explicitly

raboof commented:
an unused import only on Scala 2.13 :thinking:

johanandren commented:
I'm a bit worried that this has a too high penalty for materialization, a short run of the MaterializationBenchmark shows from pretty much no slow down to ~15% slower for some cases but I'm not sure about those numbers, need to run the bench some more and double check.

ennru reviewed (approved):
This will be very useful.
LGTM.

johanandren commented:
Re-ran benchmark today. I think the materialization hit is low enough/close enough to noise that we can accept this and deal with any specific performance problems if they arise.


```
With SourceLocation attribute
Benchmark                                                (complexity)   Mode  Cnt       Score       Error  Units   (vs baseline)
MaterializationBenchmark.flow_with_map                              1  thrpt    5  239992,282 ± 28630,943  ops/s    96%   
MaterializationBenchmark.flow_with_map                             10  thrpt    5  142910,329 ±  2849,390  ops/s    93%
MaterializationBenchmark.flow_with_map                            100  thrpt    5   38539,328 ±  1061,710  ops/s    86%
MaterializationBenchmark.graph_with_imported_flow                   1  thrpt    5  241854,339 ± 35103,036  ops/s   106%
MaterializationBenchmark.graph_with_imported_flow                  10  thrpt    5  129178,802 ±  1154,565  ops/s    96%
MaterializationBenchmark.graph_with_imported_flow                 100  thrpt    5   31139,182 ±  1062,769  ops/s    92%
MaterializationBenchmark.graph_with_junctions_gradual               1  thrpt    5  239125,225 ± 39726,150  ops/s   103%
MaterializationBenchmark.graph_with_junctions_gradual              10  thrpt    5  103421,596 ±  2063,545  ops/s    99%
MaterializationBenchmark.graph_with_junctions_gradual             100  thrpt    5    9219,574 ±   279,067  ops/s    97%
MaterializationBenchmark.graph_with_junctions_immediate             1  thrpt    5  208140,651 ± 11779,395  ops/s    98%
MaterializationBenchmark.graph_with_junctions_immediate            10  thrpt    5  156420,103 ±  5372,413  ops/s    96%
MaterializationBenchmark.graph_with_junctions_immediate           100  thrpt    5   58973,557 ±  2387,763  ops/s    98%
MaterializationBenchmark.sub_stream                                 1  thrpt    5  192419,870 ± 20621,692  ops/s    99%
MaterializationBenchmark.sub_stream                                10  thrpt    5  113372,461 ±  1416,131  ops/s    97%
MaterializationBenchmark.sub_stream                               100  thrpt    5   22558,577 ±   245,668  ops/s    90%


baseline
Benchmark                                                (complexity)   Mode  Cnt       Score       Error  Units
MaterializationBenchmark.flow_with_map                              1  thrpt    5  250048,147 ± 37096,379  ops/s
MaterializationBenchmark.flow_with_map                             10  thrpt    5  152853,149 ±  2257,300  ops/s
MaterializationBenchmark.flow_with_map                            100  thrpt    5   44561,527 ±   884,465  ops/s
MaterializationBenchmark.graph_with_imported_flow                   1  thrpt    5  237230,955 ± 18571,707  ops/s
MaterializationBenchmark.graph_with_imported_flow                  10  thrpt    5  134393,499 ±  7260,703  ops/s
MaterializationBenchmark.graph_with_imported_flow                 100  thrpt    5   33729,086 ±   500,537  ops/s
MaterializationBenchmark.graph_with_junctions_gradual               1  thrpt    5  231609,720 ± 19761,056  ops/s
MaterializationBenchmark.graph_with_junctions_gradual              10  thrpt    5  104245,973 ±   903,936  ops/s
MaterializationBenchmark.graph_with_junctions_gradual             100  thrpt    5    9551,597 ±   285,907  ops/s
MaterializationBenchmark.graph_with_junctions_immediate             1  thrpt    5  212552,446 ± 11135,513  ops/s
MaterializationBenchmark.graph_with_junctions_immediate            10  thrpt    5  161853,035 ± 11158,104  ops/s
MaterializationBenchmark.graph_with_junctions_immediate           100  thrpt    5   60006,484 ±   552,709  ops/s
MaterializationBenchmark.sub_stream                                 1  thrpt    5  195004,440 ± 17202,986  ops/s
MaterializationBenchmark.sub_stream                                10  thrpt    5  116940,618 ±  9665,168  ops/s
MaterializationBenchmark.sub_stream                               100  thrpt    5   24795,317 ±   307,664  ops/s
```

johanandren committed:
Duplicated import statement removed

johanandren commented:
@ygree this is a rehash of the other PR with stream operator source location, in this one we added a new attribute instead and updated a few `toStrings`. Cinnamon uses the `Name` attribute directly and not toString on GraphStage/GraphStageLogic, so this should be safe, right?

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1776,7 +1776,9 @@ trait FlowOps[+Out, +Mat] {
-    via(Batch(1L, ConstantFun.zeroLong, seed, aggregate).withAttributes(DefaultAttributes.conflate))
+    via(
+      Batch(1L, ConstantFun.zeroLong, seed, aggregate)
+        .withAttributes(DefaultAttributes.conflate and SourceLocation.forLambda(aggregate)))
```
how does the SourceLocation cache work? The key is the class of the lambda. Can't this be used with a `scala.Function2`?

johanandren replied:
I didn't think much about the cache lookup, just copied it from the original PR, maybe that can be a problem, I'll look into that. When benchmarking it seemed the cache does not do a huge difference for performance so maybe we should just skip it instead.

Re `Function2` that works:

```
scala> akka.stream.Attributes.SourceLocation.forLambda(akka.stream.scaladsl.Keep.left)
res1: akka.stream.Attributes.SourceLocation = SourceLocation(akka.stream.scaladsl-Materialization.scala:14)
```

patriknw replied:
That is still a lambda class. I was thinking the abnormal usage of `new Function2` from several different places. Maybe that doesn't have a source location anyway.

johanandren replied:
I think it works:

```
scala> val f = new Function2[Int, Int, Int] {
     | def apply(a: Int, b: Int): Int = a + b
     | }
f: (Int, Int) => Int = <function2>

scala> f.getClass
res0: Class[_ <: (Int, Int) => Int] = class $anon$1

scala> akka.stream.Attributes.SourceLocation.forLambda(f)
res1: akka.stream.Attributes.SourceLocation = $line6-<console>:26
```

It still gets a unique class because it is an anonymous class, however I removed the caching now so it does not matter. :)

johanandren commented:
I just realized the Materialization bench pre-constructs the blueprints and only materializes 🤦  so those numbers are not very relevant (the hit seems consistent though, guessing materialization is looking through the attributes somewhere, and the attribute list grew by one)

johanandren commented:
I take no caching needed back, several orders of magnitude slower without the cache.

```
SourceLocation attribute
Benchmark                            (complexity)   Mode  Cnt        Score       Error  Units
GraphBuilderBenchmark.flow_with_map             1  thrpt    5  2006074,377 ± 46254,486  ops/s
GraphBuilderBenchmark.flow_with_map            10  thrpt    5   440028,273 ±  7369,840  ops/s
GraphBuilderBenchmark.flow_with_map           100  thrpt    5    49191,323 ±  1024,077  ops/s
GraphBuilderBenchmark.flow_with_map          1000  thrpt    5     4864,424 ±   540,625  ops/s

SourceLocation attribute, no cache
GraphBuilderBenchmark.flow_with_map             1  thrpt    5  915,667 ± 14,071  ops/s
GraphBuilderBenchmark.flow_with_map            10  thrpt    5   91,233 ±  1,945  ops/s
GraphBuilderBenchmark.flow_with_map           100  thrpt    5    9,122 ±  0,123  ops/s
GraphBuilderBenchmark.flow_with_map          1000  thrpt    5    0,908 ±  0,039  ops/s

baseline
GraphBuilderBenchmark.flow_with_map             1  thrpt    5  2168566,233 ± 69532,288  ops/s
GraphBuilderBenchmark.flow_with_map            10  thrpt    5   487062,918 ±  3871,123  ops/s
GraphBuilderBenchmark.flow_with_map           100  thrpt    5    57233,815 ±   863,505  ops/s
GraphBuilderBenchmark.flow_with_map          1000  thrpt    5     5764,395 ±    92,741  ops/s
```

I'll verify that the class makes sense, but it is problematic with a static cache for something that could potentially be generated dynamically (lambdas closing over some instance variable) and being stuck in the cache.

johanandren committed:
Lazy creation of string, no caching.

johanandren commented:
I came up with something better. We only rarely actually use that string, so better create it when it is used and just keep a reference to the lambda in the attribute instead, since the attribute is attached to the stream operator it should be gc:d fine together with the operator when that is no longer used.

Much closer to baseline 👍 
```
[info] Benchmark                            (complexity)   Mode  Cnt        Score       Error  Units
[info] GraphBuilderBenchmark.flow_with_map             1  thrpt    5  2151397,423 ± 32125,165  ops/s
[info] GraphBuilderBenchmark.flow_with_map            10  thrpt    5   449591,545 ±  1568,294  ops/s
[info] GraphBuilderBenchmark.flow_with_map           100  thrpt    5    48872,650 ±  2825,224  ops/s
[info] GraphBuilderBenchmark.flow_with_map          1000  thrpt    5     5103,083 ±   675,149  ops/s
```

patriknw reviewed (approved):
LGTM

jrudolph commented:
:+1: Thanks @johanandren for getting this idea into a mergeable state!
